### PR TITLE
MP-116 SDK paginator

### PR DESCRIPTION
**Pull request by NickVries**
[MP-116](https://myparcelcombv.atlassian.net/browse/MP-116)

### CODE REVIEW

File: src/Collection/PromiseCollection.php [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    /**
+     * Counts the amount of resources in the collection.
+     *
+     * @return int
+     */
+    public function count()
+    {
+        if (!isset ($this->count)) {
```
Comment (Shodske) [inline src/Collection/PromiseCollection.php]: Weird space!

File: src/Collection/PromiseCollection.php [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+     *
+     * @return ResourceInterface[]
+     */
+    public function get()
+    {
+        if (isset($this->count) && $this->offset >= $this->count) {
+            return [];
+        }
+
+        $firstPage = ceil(($this->offset + 1) / 30);
```
Comment (Shodske) [inline src/Collection/PromiseCollection.php]: You should add a comment of what's going on here. I can imagine that someone looking at this code has no idea what the 2 pages are.

File: src/Collection/PromiseCollection.php [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        return $this;
+    }
+
+    /**
+     * Sets the amount of resources to be retrieved by get().
+     * Max and default limit is 30.
+     *
+     * @param int $limit
+     * @return $this
+     * @throws \Exception
```
Comment (Shodske) [inline src/Collection/PromiseCollection.php]: This method doesn't throw any exception.

File: src/MyParcelComApi.php [hunk 4]
```diff
@@ -505,11 +501,28 @@ protected function getResourcesPromise($uri, $ttl = self::TTL_10MIN)
...
+        return new PromiseCollection(
+            function ($pageNumber, $pageSize) use ($uri, $ttl) {
+                $url = (new UrlBuilder($uri))->addQuery([
+                    'page[number]' => $pageNumber,
+                    'page[size]'   => $pageSize,
+                ])->getUrl();
+
+                return $this->doRequest($url, 'get', [], [], $ttl);
+            }, function ($data) {
+            return $this->jsonToResources($data);
```
Comment (Shodske) [inline src/MyParcelComApi.php]: Here seems to be some weird indenting.

File: src/MyParcelComApi.php [hunk 5]
```diff
@@ -505,11 +501,28 @@ protected function getResourcesPromise($uri, $ttl = self::TTL_10MIN)
...
+        });
+    }
+
     /**
      * {@inheritdoc}
      */
-    public function doRequest($uri, $method = 'get', array $body = [], array $headers = [], $ttl = self::TTL_10MIN)
-    {
+    public
```
Comment (Shodske) [inline src/MyParcelComApi.php]: Public on it's own line?
All other functions seem to have this issue, please correct!

File: src/MyParcelComApi.php [hunk 6]
```diff
@@ -505,11 +501,28 @@ protected function getResourcesPromise($uri, $ttl = self::TTL_10MIN)
             });
     }
 
+    protected function getResourceCollection($uri, $ttl = self::TTL_10MIN)
```
Comment (Shodske) [inline src/MyParcelComApi.php]: Docblock

File: src/MyParcelComApiInterface.php [hunk 7]
```diff
@@ -44,15 +44,15 @@ public function authenticate(AuthenticatorInterface $authenticator);
      *
      * @param string|null $countryCode
      * @param string|null $regionCode
-     * @return RegionInterface[]
+     * @return PromiseCollection
```
Comment (Shodske) [inline src/MyParcelComApiInterface.php]: This should be `CollectionInterface` everywhere in this interface.

File: tests/Feature/MyParcelComApiTest.php [hunk 8]
```diff
@@ -230,45 +231,46 @@ public function testGetRegions()
...
     /** @test */
     public function testGetGbRegions()
     {
         $regions = $this->api->getRegions('GB');
 
-        $this->assertInternalType('array', $regions);
-        $this->assertCount(5, $regions);
-        array_walk($regions, function ($region) {
+        $this->assertInstanceOf(CollectionInterface::class, $regions);
+        $this->assertEquals(9, $regions->count());
```
Comment (Shodske) [inline tests/Feature/MyParcelComApiTest.php]: Did you adjust the number of regions in the stubs?

File: tests/Feature/MyParcelComApiTest.php [hunk 9]
```diff
@@ -299,37 +301,42 @@ public function testGetServicesForCarrier()
...
     public function testGetShipments()
     {
         $shipments = $this->api->getShipments();
 
-        $this->assertInternalType('array', $shipments);
-        array_walk($shipments, function ($shipment) {
+//        $this->assertInternalType('array', $shipments);
+//        array_walk($shipments, function ($shipment) {
+//            $this->assertInstanceOf(ShipmentInterface::class, $shipment);
+//        });
```
Comment (Shodske) [inline tests/Feature/MyParcelComApiTest.php]: Remove comments and fix the test.

File: src/Collection/ArrayCollection.php [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+    /** @var int */
+    protected $currentResourceNumber = 0;
+
+    /**
+     * ArrayCollection constructor.
+     *
+     * @param ResourceInterface[] $resources
+     */
+    public function __construct($resources)
```
Comment (Shodske) [inline src/Collection/ArrayCollection.php]: Typehint `array`

File: tests/Feature/MyParcelComApiTest.php (hunk 8 above)
Comment (NickVries) [reply tests/Feature/MyParcelComApiTest.php]: Retrieval of the regions is now filtered in the request not after request. I added a new stub which is just raw data returned by the api when doing that request.

File: src/Collection/ArrayCollection.php [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    protected $resources;
+
+    /** @var int */
+    protected $offset = 0;
+
+    /** @var int */
+    protected $limit = 30;
+
+    /** @var int */
+    protected $count;
```
Comment (Shodske) [inline src/Collection/ArrayCollection.php]: You're not using this property, remove it.

File: src/MyParcelComApi.php [hunk 12]
```diff
@@ -166,37 +168,40 @@ public function getPickUpDropOffLocations(
...
+        $pudoLocations = [];
 
         foreach ($carriers as $carrier) {
-            $carrierUri = str_replace('{carrier_id}', $carrier->getId(), $uri);
+            $carrierUri = str_replace('{carrier_id}', $carrier->getId(), $uri->getUrl());
 
             // These resources can be stored for a week.
             $promise = $this->getResourcesPromise($carrierUri, self::TTL_WEEK);
             if ($specificCarrier) {
                 return $promise->wait();
```
Comment (Shodske) [inline src/MyParcelComApi.php]: Should these not be wrapped in an `ArrayCollection`?

File: tests/Traits/MocksApiCommunication.php [hunk 13]
```diff
@@ -36,6 +36,8 @@ protected function getClientMock()
                     $filePath = preg_replace('/.json/', '.txt', $filePath);
                 }
 
+                $filePath = preg_replace('/\-page\-(number|size)\-\-[0-9]*/', '', $filePath);
```
Comment (Shodske) [inline tests/Traits/MocksApiCommunication.php]: Removing page number and size from the url means we cannot test pagination anymore, this should still be here.

**Pull request merged**